Refactor and speed up `cirq.transformers.stratify`

perlinm opened the pull request:
This PR refactors `cirq.transformers.stratify.stratified_circuit`, making it considerably faster.

I have not contributed to `cirq` before, so I apologize if I am not doing things the "correct" way.  Feedback and suggestions much appreciated :slightly_smiling_face:.

Some comments:
1. If there are **no** `ops_to_ignore`, for a given input (circuit + categories), the output of the new stratifier is identical to the old.
2. Ignored ops are treated a bit differently, though I think the new treatment is consistent with the previous docstring.  In the old stratifier, (a) each moment in the circuit is stratified into `num_categories` strata, and then (b) every operation is moved into the earliest stratum possible that matches its category.  Ignored ops are not moved into earlier strata at step (b), which can block non-ignored ops from moving into earlier moments.  The new stratifier moves ignored ops into earlier moments as well, which can result in shorter circuits.  The ignored ops are still "ignored" in the sense that their relative order is unaffected by stratification.  If one ignored op preceded another ignored op before stratification, that will still be true after stratification, and two ignored ops will be in the same moment after stratification if and only if they were in the same moment before stratification.
3. I claim that thew new stratifier is faster, but I am not sure how to run a proper regression test that others can conveniently use to verify my claim.  One simple test that I can run locally is
```python
import cirq
import numpy as np
import time

np.random.seed(0)
circuit = cirq.testing.random_circuit(qubits=4, n_moments=100, op_density=1)
categories = [lambda op: len(op.qubits) == 1, cirq.CNOT]  # single-qubit ops, and CNOTs

start = time.time()
stratified_circuit = cirq.stratified_circuit(circuit, categories=categories)
print("time:", time.time() - start)
```
On my laptop, this test takes ~0.4 seconds with the old stratifier, and ~0.03 seconds with the new stratifier.

Edit: it looks like this PR is failing some tests, but at the time of writing all failures appear to be some issue with installing `npm`.  I am not sure what to do about this...

perlinm commented:
I actually have another stratifier that, among other advantages, (1) is another ~10x faster on the example above (so overall ~100x faster than the current stratifier in Cirq), and (2) generally yields shorter stratified circuits.  However, this other stratifier is more complex than the current one in Cirq, and the corresponding PR would be considerably larger.  There is also the question of whether Cirq would want to keep both stratifiers, or only one.

My thinking was to first get through a "minor" stratifier improvement in this PR, and then open a separate, larger PR for the other stratifier.  However, maybe it makes more sense to add the other stratifier to this PR as well.  Thoughts?

tanujkhattar commented on cirq-core/cirq/transformers/stratify.py:
```diff
@@ -116,43 +113,101 @@ def _stratify_circuit(
-        stratified_circuit.batch_remove(batch_removals)
-        stratified_circuit.batch_insert_into(batch_inserts)
-    return drop_empty_moments.drop_empty_moments(stratified_circuit)
+            ignored_op = any(tag in op.tags for tag in context.tags_to_ignore)
```
Move this just above the `if ignored_op:` check at line 150, it reduces the cognitive overhead for the reader to track how we are dealing ignored ops differently inside the loop.

tanujkhattar commented on cirq-core/cirq/transformers/stratify.py:
```diff
@@ -116,43 +113,101 @@ def _stratify_circuit(
+            # Get the earliest time index that can accomodate this op, based on its qubits.
+            min_time_index_for_op = max(qubit_time_index[qubit] + 1 for qubit in op.qubits)
+            if ignored_op:
+                min_time_index_for_op = max(min_time_index_for_op, min_time_index_for_ignored_op)
```
Similarly, move inside the single `if ignored_op:` condition at line 150.

tanujkhattar commented on cirq-core/cirq/transformers/stratify.py:
```diff
@@ -116,43 +113,101 @@ def _stratify_circuit(
+    control_time_index: Dict['cirq.MeasurementKey', int] = {}
+
+    # The minimum time index for operations with a tag in context.tags_to_ignore.
+    min_time_index_for_ignored_op = 0
```
Consider renaming `min_time_index_for_ignored_op` to `last_ignored_ops_moment_index` because we are essentially keeping track of the latest moment that contains only ignored ops so that we can insert the next ignores ops batch after that time. 

Also, compute `min_time_index_for_ignored_op` as just the `max(min_time_index_for_ignored_op, time_index)` in the inner loop and get rid of the check `min_time_index_for_op = max(min_time_index_for_op, min_time_index_for_ignored_op)`. Once you have `min_time_index_for_ignored_op` in the inner loop, you can take the max with `last_ignored_ops_moment_index` once outside before assigning `op_time_indices` to `ignored_ops` and then update the `last_ignored_ops_moment_index` with the current index. 


In general, I think the current logic to take care of ignored ops is a bit convoluted and it's hard to reason about it's correctness. My proposed changes should make it easier for a reader to quickly understand how we are handling the `ignored_ops`.

perlinm replied:
Thanks for these suggestions for how to clean up `ignored_op` handing.  I'm not sure I implemented them exactly as requested, but the new version is certainly much cleaner.  What do you think?

tanujkhattar reviewed (changes requested):
Thanks a lot for opening the PR! The changes look great overall. 

This aligns well with our general roadmap of making Cirq more performant. I have left a round of comments highlighting a bug and with some suggests to restructure the code. 

Overall, this is looking good. We can merge once the comments have been addressed.

tanujkhattar reviewed (commented):
> My thinking was to first get through a "minor" stratifier improvement in this PR, and then open a separate, larger PR for the other stratifier

I agree with this. Let's get this PR in and then we can open a separate PR to either replace this or add a new stratifier.

perlinm commented:
Thanks for the detailed review and suggested changes!


I agree that it's quite difficult to keep track of how `ignored_ops` are handled at the moment.  Dealing with `ignored_ops` (and measurement keys) was most of the work in this PR...  Hopefully your suggestions clean this up somewhat.

Anyways, I'll try to get to your suggested changes later this week.

perlinm commented on cirq-core/cirq/transformers/stratify.py:
```diff
@@ -116,43 +113,101 @@ def _stratify_circuit(
+                min_time_index_for_op = max(min_time_index_for_op, time_index + 1)
+            for key in protocols.measurement_key_objs(op) & control_time_index.keys():
+                time_index = control_time_index[key]
+                min_time_index_for_op = max(min_time_index_for_op, time_index + 1)
```
I factored out the logic for finding the earliest available/accommodating moment, and tentatively placed it in `circuits.circuit.get_earliest_accommodating_moment_index`.  Suggestions welcome for where this function should be moved.

There are currently two related points that I am dissatisfied with:
1. `circuits/circuit.py` currently has to construct `mop_qubits/mkeys/ckeys` twice: (a) in [`_load_contents_with_earliest_strategy`](https://github.com/perlinm/Cirq/blob/090beb63e805a9b7e7a23e3014928940f9161853/cirq-core/cirq/circuits/circuit.py#L1782-L1784), and (b) in [`get_earliest_accomodating_moment_index`](https://github.com/perlinm/Cirq/blob/090beb63e805a9b7e7a23e3014928940f9161853/cirq-core/cirq/circuits/circuit.py#L2760-L2762).
2. Despite using the same method/logic for finding the earliest accommodating moment, `circuits/circuit.py` and `transformers/stratify.py` have to independently update the `qubit/mkey/ckey_index` dictionaries.

Fundamentally, both of these issues stem from the fact that `_load_contents_with_earliest_strategy` has special logic for what to do with `Moment`s rather than `Operation`s, which gets tangled with the logic of identifying an operation's `qubits/mkeys/ckeys`, finding the earliest accommodating moment, and updating the `qubit/mkey/ckey --> index` dictionaries.

Ideas for how to fix these issues?

perlinm replied:
Two possible resolutions (though there may be better ones):
- Make `get_earliest_accommodating_moment_index` accept a `Union[cirq.Operation, cirq.Moment]`, and for moments return something like `max(*qubit_indices.values(), *mkey_indices.values(), *ckey_indices.values())`.  This would make the code relatively "clean", but it would require  to compute a maximum every time (whereas the current implementation keeps track of this maximum "manually").
- Again accept a `cirq.Moment` as an input, but also accept an optional `circuit_length: Optional[int]` argument, and have `get_earliest_accommodating_moment_index` return that `if isinstance(mop, cirq.Moment) and circuit_length is not None`.  This avoids recalculating the circuit length by maximizing over indices, but it is potentially "dangerous", since he method would happily return an invalid circuit length.  On the other hand, garbage in garbage out so maybe that's okay?

perlinm commented:
@tanujkhattar I believe that I have addressed all comments/requests.  Please let me know of any other suggestions for what to change in this PR!

For example: I suspect that there is a better place for [this new method](https://github.com/perlinm/Cirq/blob/d567650012bcfd76bdde3e0eb720a001c5319777/cirq-core/cirq/circuits/circuit.py#L2727) to live.

tanujkhattar reviewed (approved):
This looks great! Let's merge this and open a separate issue to track what's the best place to put the free function `get_earliest_accommodating_moment_index`.


Thank you for the contribution and apologies for the long wait! Let's merge this PR and keep the contributions coming! :)

perlinm commented:
Thank you!  Also thanks to @vtomole for some brief consultation on how to structure some of the changes in this PR :slightly_smiling_face: 

It looks like some of the continuous integration tests are still failing, but that appears to be due to code that was not touched by this PR.  What should I do about it?
